File: src/hooks.rs
use std::collections::BTreeMap;
use std::time::Duration;

use serde_json::{Map, Value};

use crate::exec::{run, RunOptions};

pub type Hooks = BTreeMap<String, Vec<String>>;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HookEvent {
    TaskStart,
    StageChange,
    Attention,
    TaskNeedsInput,
    TaskBlocked,
    TaskRetrying,
    TaskDone,
    LoopIdle,
}

impl HookEvent {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::TaskStart => "task.start",
            Self::StageChange => "stage.change",
            Self::Attention => "attention",
            Self::TaskNeedsInput => "task.needs_input",
            Self::TaskBlocked => "task.blocked",
            Self::TaskRetrying => "task.retrying",
            Self::TaskDone => "task.done",
            Self::LoopIdle => "loop.idle",
        }
    }
}

const HOOK_TIMEOUT: Duration = Duration::from_secs(10);

pub fn flat_env(payload: &Map<String, Value>) -> BTreeMap<String, String> {
    let mut out = BTreeMap::new();
    for (key, value) in payload {
        if value.is_null() {
            continue;
        }
        let value = value
            .as_str()
            .map(ToOwned::to_owned)
            .unwrap_or_else(|| value.to_string());
        out.insert(format!("FACTORY_{}", key.to_uppercase()), value);
    }
    out
}

pub fn emit(root: &str, hooks: &Hooks, event: HookEvent, payload: Map<String, Value>) {
    let Some(commands) = hooks.get(event.as_str()) else {
        return;
    };
    if commands.is_empty() {
        return;
    }

    let mut stdin = Map::new();
    stdin.insert(
        "event".to_string(),
        Value::String(event.as_str().to_string()),
    );
    stdin.extend(payload.clone());
    let stdin = Value::Object(stdin).to_string();

    let mut env: BTreeMap<String, String> = std::env::vars().collect();
    env.insert("FACTORY_EVENT".to_string(), event.as_str().to_string());
    env.insert("FACTORY_ROOT".to_string(), root.to_string());
    env.extend(flat_env(&payload));

    for command in commands {
        match run(
            &["bash".to_string(), "-lc".to_string(), command.clone()],
            &RunOptions {
                cwd: root.to_string(),
                stdin: Some(stdin.clone()),
                env: Some(env.clone()),
                timeout: Some(HOOK_TIMEOUT),
                ..RunOptions::default()
            },
        ) {
            Ok(result) if result.code == 0 => {}
            Ok(result) => eprintln!(
                "warning: hook {} ({}) exited {}",
                event.as_str(),
                command,
                result.code
            ),
            Err(err) => eprintln!(
                "warning: hook {} ({}) failed: {}",
                event.as_str(),
                command,
                err
            ),
        }
    }
}

#[cfg(test)]
mod tests {
    use std::fs;

    use serde_json::json;
    use tempfile::TempDir;

    use super::*;

    #[test]
    fn flat_env_skips_null_and_json_encodes_non_strings() {
        let mut payload = Map::new();
        payload.insert("stage".to_string(), Value::String("plan".to_string()));
        payload.insert("active".to_string(), Value::Bool(true));
        payload.insert("skip".to_string(), Value::Null);

        let env = flat_env(&payload);
        assert_eq!(env.get("FACTORY_STAGE"), Some(&"plan".to_string()));
        assert_eq!(env.get("FACTORY_ACTIVE"), Some(&"true".to_string()));
        assert!(!env.contains_key("FACTORY_SKIP"));
    }

    #[test]
    fn emit_delivers_stdin_and_env() {
        let dir = TempDir::new().unwrap();
        let out = dir.path().join("hook.out");
        let command = format!(
            "printf '%s\\n' \"$FACTORY_EVENT:$FACTORY_STAGE:$FACTORY_ACTIVE\" > {}; cat >> {}",
            out.display(),
            out.display()
        );
        let mut hooks = Hooks::new();
        hooks.insert("stage.change".to_string(), vec![command]);

        let mut payload = Map::new();
        payload.insert("stage".to_string(), json!("plan"));
        payload.insert("active".to_string(), json!(true));
        emit(
            dir.path().to_str().unwrap(),
            &hooks,
            HookEvent::StageChange,
            payload,
        );

        let text = fs::read_to_string(out).unwrap();
        assert!(text.contains("stage.change:plan:true"));
        assert!(text.contains(r#""event":"stage.change""#));
        assert!(text.contains(r#""stage":"plan""#));
        assert!(text.contains(r#""active":true"#));
    }
}

File: src/hooks.ts
import { run } from './exec.ts'
import { log } from './log.ts'

// Lifecycle hooks: at each event, factory runs the shell commands configured for
// it (in `.factory.json` "hooks"), letting the surrounding environment react —
// drive a tmux tab, send a notification, log to a dashboard — WITHOUT factory
// knowing anything about that environment. This is what lets factory be a
// standalone program: the only built-in integration (tmux) moves out into a hook.
//
// Best-effort by design: a hook's failure or timeout is logged and never affects
// the task (cosmetic integration must not crash the loop). One-directional —
// hooks observe, they don't gate; their output and exit code don't alter flow.

export type Hooks = Record<string, string[]>

export type HookEvent =
  | 'task.start' // a task begins processing
  | 'stage.change' // the active stage changed (drives the tmux window name)
  | 'attention' // attention state changed
  | 'task.needs_input' // paused for the human
  | 'task.blocked' // escalated to blocked
  | 'task.retrying' // set aside for an auto-retry
  | 'task.decomposed' // delegated into an ordered child-workstream chain
  | 'task.done' // completed (committed / shipped)
  | 'task.stale' // parked (needs-input/blocked) past the staleness window
  | 'loop.idle' // queue drained, loop waiting

const HOOK_TIMEOUT_MS = 10_000

// Flatten a payload into FACTORY_* env vars, so trivial shell hooks need no JSON
// parsing (e.g. `tmux rename-window "$FACTORY_STAGE"`). Strings pass through;
// everything else is JSON-encoded.
function flatEnv(payload: Record<string, unknown>): Record<string, string> {
  const out: Record<string, string> = {}
  for (const [key, value] of Object.entries(payload)) {
    if (value === null || value === undefined) {
      continue
    }
    out[`FACTORY_${key.toUpperCase()}`] = typeof value === 'string' ? value : JSON.stringify(value)
  }
  return out
}

// Fire a lifecycle event to its configured hooks. Payload is delivered both as a
// JSON object on stdin and as flat FACTORY_* env vars; cwd is the worktree root,
// and the child inherits the process env (so $TMUX_PANE etc. are available).
export async function emit(
  root: string,
  hooks: Hooks,
  event: HookEvent,
  payload: Record<string, unknown> = {}
): Promise<void> {
  const commands = hooks[event]
  if (!commands || commands.length === 0) {
    return
  }
  const stdin = JSON.stringify({ event, ...payload })
  const env = { ...process.env, FACTORY_EVENT: event, FACTORY_ROOT: root, ...flatEnv(payload) }
  for (const command of commands) {
    try {
      const res = await run(['bash', '-lc', command], {
        cwd: root,
        stdin,
        env,
        timeout: HOOK_TIMEOUT_MS,
      })
      if (res.code !== 0) {
        log.warn(`hook ${event} (${command}) exited ${res.code}`)
      }
    } catch (err) {
      log.warn(`hook ${event} (${command}) failed: ${err instanceof Error ? err.message : err}`)
    }
  }
}
